lib/imagery/manage_signatures.c

```diff
@@ -0,0 +1,254 @@
+
+#include <grass/gis.h>
+#include <grass/imagery.h>
+#include <grass/glocale.h>
+
+static int list_by_type(int, const char *, int, char ***);
+
+/*!
+ * \brief Remove a signature file
+ * 
+ * If removal fails, prints a warning and returns 1.
+ * It is safe to pass fully qualified names.
+ * 
+ * \param type SIGFILE_TYPE_ signature type
+ * \param name of signature to remove
+ * \return 0 on success
+ * \return 1 on failure
+ */
+int I_signatures_remove(int type, const char *name)
+{
+    char xname[GNAME_MAX], xmapset[GMAPSET_MAX];
+    char selem[GNAME_MAX];
+
+    G_debug(1, "I_signatures_remove(%d, %s);", type, name);
+
+    /* Remove only if file is in the current mapset */
+    if (G_name_is_fully_qualified(name, xname, xmapset) &&
+        strcmp(xmapset, G_mapset()) != 0) {
+        G_warning(_("%s is not in the current mapset (%s)"), name,
+                  G_mapset());
+        return 1;
+    }
+    if (I_find_signature2(type, name, G_mapset())) {
+        if (type == SIGFILE_TYPE_SIG)
+            sprintf(selem, "signatures%csig", HOST_DIRSEP);
+        else if (type == SIGFILE_TYPE_SIGSET) {
+            sprintf(selem, "signatures%csigset", HOST_DIRSEP);
+        }
+        if (G_remove(selem, name) == 1) {
+            G_verbose_message(_("%s removed"), name);
+            return 0;
+        }
+        G_warning(_("Unable to remove %s signature"), name);
+    }
+    else
+        G_warning(_("%s is missing"), name);
+    return 1;
+}
+
+/*!
+ * \brief Copy a signature file
+ *
+ * If copy fails, prints warning messages and returns 1.
+ * It is safe to pass fully qualified names.
+ *
+ * \param type SIGFILE_TYPE_ signature type
+ * \param name of old signature
+ * \param mapset of old signature
+ * \param name of new signature
+ * \return 0 on success
+ * \return 1 on failure
+ */
+int I_signatures_copy(int type, const char *old_name, const char *old_mapset,
+                      const char *new_name)
+{
+    char tname[GNAME_MAX], tmapset[GMAPSET_MAX];
+    char selem[GNAME_MAX];
+    const char *smapset;
+    char old_path[GPATH_MAX], new_path[GPATH_MAX];
+
+    G_debug(1, "I_signatures_copy(%d, %s@%s, %s);", type, old_name,
+            old_mapset, new_name);
+
+    /* Copy only if mapset of new name is the current mapset */
+    if (G_name_is_fully_qualified(new_name, tname, tmapset)) {
+        if (strcmp(tmapset, G_mapset()) != 0) {
+            G_warning(_("%s is not in the current mapset (%s)"), new_name,
+                      G_mapset());
+            return 1;
+        }
+    }
+    else
+        strcat(tname, new_name);
+
+    smapset = I_find_signature2(type, old_name, old_mapset);
+    if (!smapset) {
+        G_warning(_("%s is missing"), old_name);
+        return 1;
+    }
+
+    if (type == SIGFILE_TYPE_SIG)
+        sprintf(selem, "signatures%csig", HOST_DIRSEP);
+    else if (type == SIGFILE_TYPE_SIGSET) {
+        sprintf(selem, "signatures%csigset", HOST_DIRSEP);
+    }
+
+    G_make_mapset_element(selem);
```

Comment:
I reworked all management functions to not use sprintf but use G_.._misc instead (still keeping the "name" is "element" and "element" is "signature type" confusion in place).

Reply:
I could go with Vlib/open approach – nobody even bothers to use HOST_DIRSEP there :D
https://github.com/OSGeo/grass/blob/b77880d09fa2f284b1be4f5d7732aa03d711934f/lib/vector/Vlib/open.c#L1455

1. I could create an internal I__get_sigfile_path to create "signature" + HOST_DIRSEP + "sig"
2. I could pass "element/subelement" as element to non-_misc functions thus eliminating confusion in naming & not needing to add new G_open|find functions.

Reply:
> I could go with Vlib/open approach – nobody even bothers to use HOST_DIRSEP there

I don't think HOST_DIRSEP is that important since `/` is a universally supported dir sep, although it would be nice to show users path with preferred dir seps or at least not show paths with mixed seps. I think the important part is to concentrate all path tricks in one place, preferably in lib gis and not have then intermingled with other code, so that they can be changed or fixed later easily.

If `Vect__get_path` in `Vlib/open.c` would be the only place in Vlib which has `/` hardcoded that would be fine. It is _the_ place to get path and it is a path, not element name or something abstract like that.

>     1. I could create an internal I__get_sigfile_path to create "signature" + HOST_DIRSEP + "sig"

I'm not sure what exactly this function would do. You ultimately need name there, right? Full path perhaps? Also, you are making the element name to be signature/sig, so your functions may need to accept that to make it convenient, no?

>     2. I could pass "element/subelement" as element to non-_misc functions thus eliminating confusion in naming & not needing to add new G_open|find functions.

This would come back to us when we try to change the raster file to `raster/name/{fcell,hist,range,...}`. However, it may work well if you wrap that into another function which will be the "element/subelement"-aware function which we can later change if needed.

Reply:
I implemented a version of workaround.